### PR TITLE
Remove NumPy <2 pin, remove proj pin

### PR DESCRIPTION
**Pull request by seberg**
This PR removes the NumPy<2 pin which is expected to work for
RAPIDS projects once CuPy 13.3.0 is released (CuPy 13.2.0 had
some issues preventing the use with NumPy 2).

Also contributes to https://github.com/rapidsai/build-planning/issues/56 by removing the pin on `proj` (which allows this project to work with newer `fmt` / `spdlog`, in sync with conda-forge).

### CODE REVIEW

**Comment by bdice** [comment]
I merged in the upstream to see if CI is failing in the same way we see here: https://github.com/rapidsai/docker/pull/709#issuecomment-2302229715

**Comment by jakirkham** [comment]
Closing and reopening to restart CI now that cuDF is done: https://github.com/rapidsai/cudf/pull/16300

**Comment by jakirkham** [comment]
Seeing some unexpected and unrelated failures that appear to be due to a cuDF change

Running CI in a no change PR to see if this reproduces there: https://github.com/rapidsai/cuspatial/pull/1444

**Comment by jakirkham** [comment]
On the Conda side, it looks like we are still seeing [NumPy 1 in CI]( https://github.com/rapidsai/cuspatial/actions/runs/10536209879/job/29196463546?pr=1441#step:7:465 ). 

```
numpy                     1.26.4          py311h64a7726_0    conda-forge
```

Think we are missing another dependency. Will need to take a closer look

**Comment by jakirkham** [comment]
Took all of the dependencies on CI (except RAPIDS ones) and tried to create a Conda environment with them locally on Linux ARM (though there should be no difference between x86_64 in this regard). It created the environment using NumPy 2

Didn't see anything obviously missing in this repo

The other RAPIDS dependencies are all migrated

Not sure what we are missing. May need to dig deeper later

**Comment by jameslamb** [comment]
> Edit: Seeing the same error over there ( https://github.com/rapidsai/cuspatial/pull/1444#issuecomment-2308151120 ). Following up separately

Strongly suspect this is a result of the changes from https://github.com/rapidsai/cudf/pull/15483.

I'll work on adapting `cuspatial` to account for those, but going to wait just a bit until https://github.com/rapidsai/cudf/pull/16640 is merged.

**Comment by jameslamb** [comment]
#1447 is merged and I've pulled in those changes here

**Comment by jakirkham** [comment]
Ok think I understand what is happening with Conda now, OSMnx 1.9.4 (currently latest) [pins to NumPy 1]( https://github.com/gboeing/osmnx/blob/715d704b4b93bd29a4d6782923194f4f734658b2/pyproject.toml#L30 )

IIUC OSMnx 2.0 will [relax this constraint]( https://github.com/gboeing/osmnx/blob/94b7e7358e88a68ef4f07344c8d03c0da140f775/pyproject.toml#L29 )

I don't think there is a lot we can do about that

Looking at [the code here]( https://github.com/search?q=repo%3Arapidsai%2Fcuspatial%20osmnx&type=code ), it appears the cuSpatial library doesn't actually use OSMnx. Only the notebooks do

Perhaps there is a way to move that dependency out of the main CI paths so we can test with NumPy 2 here (ignoring notebooks which will have to stay on NumPy 1)

Think if we can do that, we can move forward

**Comment by jakirkham** [comment]
Should add another piece is `rasterio` needs a sufficiently recent package build to pickup up NumPy 2 support: https://github.com/conda-forge/rasterio-feedstock/pull/299

This can only happen if `proj` `9.4` is installed. However we were constraining `proj` to `9.3`. James proposed relaxing that in PR ( https://github.com/rapidsai/cuspatial/pull/1449 ), which is now merged

While `rasterio` is not a direct dependency, it does get pulled in through other dependencies we have in the spatial packages. So James fix should also help relax pinnings for NumPy 2 to more easily be picked up

**Review by jameslamb — CHANGES_REQUESTED** [review]
I just checked, and looks like the `conda-python-tests` jobs are still pulling in `numpy==1.26.4` at runtime:

https://github.com/rapidsai/cuspatial/actions/runs/10626040400/job/29457392459?pr=1441#step:7:468

`wheel-tests-cuspatial` are pulling in `numpy==2.0.2`

https://github.com/rapidsai/cuspatial/actions/runs/10626040400/job/29457234502?pr=1441#step:8:1006

`wheel-tests-cuproj` also pulling in `numpy==2.0.2`

https://github.com/rapidsai/cuspatial/actions/runs/10626040400/job/29457234993?pr=1441#step:8:1009

This PR is up to date with latest `branch-24.10` (so includes the `proj` pin removal from #1449 and other packaging changes). I think we need to figure out what else is holding `numpy` under 2.0. @jakirkham @seberg I'm going to push a commit here changing these pins to `numpy>=2.1.0,<3.0a0` just to hopefully get more direct information about this via solver errors.

**Comment by jameslamb** [comment]
In the wheel builds, looks to me like `pip` is backtracking on `cupy-cuda12x`

```text
Collecting cupy-cuda12x>=12.0.0 (from cudf-cu12==24.10.*,>=0.0.0a0->cuspatial-cu12==24.10.0a43->cuspatial-cu12==24.10.0a43)
  Downloading cupy_cuda12x-13.2.0-cp310-cp310-manylinux2014_x86_64.whl.metadata (2.7 kB)
INFO: pip is still looking at multiple versions of numba to determine which version is compatible with other requirements. This could take a while.
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. See https://pip.pypa.io/warnings/backtracking for guidance. If you want to abort this run, press Ctrl + C.
  Downloading cupy_cuda12x-13.1.0-cp310-cp310-manylinux2014_x86_64.whl.metadata (2.7 kB)
INFO: pip is looking at multiple versions of cupy-cuda12x to determine which version is compatible with other requirements. This could take a while.
  Downloading cupy_cuda12x-13.0.0-cp310-cp310-manylinux2014_x86_64.whl.metadata (2.6 kB)
  Downloading cupy_cuda12x-12.3.0-cp310-cp310-manylinux2014_x86_64.whl.metadata (2.7 kB)
  Downloading cupy_cuda12x-12.2.0-cp310-cp310-manylinux2014_x86_64.whl.metadata (2.6 kB)
  Downloading cupy_cuda12x-12.1.0-cp310-cp310-manylinux2014_x86_64.whl.metadata (2.6 kB)
  Downloading cupy_cuda12x-12.0.0-cp310-cp310-manylinux2014_x86_64.whl.metadata (2.6 kB)
Collecting geopandas>=0.11.0 (from cuspatial-cu12==24.10.0a43->cuspatial-cu12==24.10.0a43)
  Downloading geopandas-1.0.0-py3-none-any.whl.metadata (2.2 kB)
INFO: pip is still looking at multiple versions of cupy-cuda12x to determine which version is compatible with other requirements. This could take a while.
...
ERROR: Cannot install cudf-cu12, cuspatial-cu12 and cuspatial-cu12[test]==24.10.0a43 because these package versions have conflicting dependencies.
```

Seems like it isn't even considering `cupy-cuda12x>=13.2.0`. 

Just pushed https://github.com/rapidsai/cuspatial/pull/1441/commits/c1f58458154195bed323292e1626fa46db16c8ab bumping to `cupy>=13.2.0`.... let's see if it helps us get more information from the resolvers.

**Comment by seberg** [comment]
Is the reason that something is pinning llvmlite to:
`Collecting llvmlite<0.43,>=0.42.0dev0` at https://github.com/rapidsai/cuspatial/actions/runs/10635279113/job/29485001500?pr=1441#step:8:833 ?
But it isn't clear to me what is doing that.

File: conda/environments/all_cuda-118_arch-x86_64.yaml [hunk 1]
```diff
@@ -13,7 +13,7 @@ dependencies:
 - cudatoolkit
 - cudf==24.10.*,>=0.0.0a0
 - cuml==24.10.*,>=0.0.0a0
-- cupy>=12.0.0
+- cupy>=13.2.0
```
Comment (seberg) [inline conda/environments/all_cuda-118_arch-x86_64.yaml:16]:
```suggestion
- cupy>=13.2.0
- llvmlite>=0.43.0
```
Let's see if this gives us information (if my llvmlite guess was right)...

File: conda/recipes/cuspatial/meta.yaml [hunk 2]
```diff
@@ -77,7 +77,7 @@ requirements:
     - cudf ={{ minor_version }}
     - pylibcudf ={{ minor_version }}
     - geopandas >=0.11.0
-    - numpy >=1.23,<2.0a0
+    - numpy >=2.1.0,<3.0a0
```
Comment (seberg) [inline conda/recipes/cuspatial/meta.yaml:80]:
```suggestion
    - numpy >=2.0.0,<3.0a0
```
Well, that was silly of me. We can't use 2.1 yet (due to numba), so this needed adjustment.  The wheel tests seem to be fine, let's hope that this fixes the conda ones.
(After which, need to undo any other shenanigans.)

**Comment by seberg** [comment]
Tried removing all new things and rebasing once more...  I am confuse because it seems like there is something that just doesn't resolve in conda, but I am not sure what.

**Comment by bdice** [comment]
@seberg This is blocked because we're pinning to an old `proj`. See https://github.com/rapidsai/cuspatial/blob/2a30f5f3c4153a65af2a57756661cc7e96634533/dependencies.yaml#L183

This will be unblocked once we migrate to a newer fmt/spdlog. See https://github.com/rapidsai/build-planning/issues/56. cc: @jameslamb

**Comment by jameslamb** [comment]
Ah thanks for making that connection @bdice ! Helpful to know, for prioritization, that the `fmt` / `spdlog` stuff is also necessary to unblock supporting newer NumPy.

**Comment by jameslamb** [comment]
🎉 🎉 🎉 

```text
...
fmt                       11.0.2               h70be974_0    conda-forge
...
numpy                     2.0.2           py311he9aa9f1_0    conda-forge
...
proj                      9.5.0                h07e4b22_0    conda-forge
...
spdlog                    1.14.1               h9d9cc24_1    conda-forge
...
tiledb                    2.26.1               hf11d815_0    conda-forge
...
```

([build-python-tests link](https://github.com/rapidsai/cuspatial/actions/runs/10933768300/job/30353133682?pr=1441))

File: conda/recipes/cuspatial/meta.yaml [hunk 3]
```diff
@@ -77,7 +77,7 @@ requirements:
     - cudf ={{ minor_version }}
     - pylibcudf ={{ minor_version }}
     - geopandas >=1.0.0
-    - numpy >=1.23,<2.0a0
+    - numpy  >=1.23,<3.0a0
```
Comment (bdice) [inline conda/recipes/cuspatial/meta.yaml:80]: Revert extra space.
```suggestion
    - numpy >=1.23,<3.0a0
```

**Review by bdice — APPROVED** [review]

**Comment by jameslamb** [comment]
Ran CI again after removing all testing/debugging changes... still getting `numpy` 2.x!

I think this can be merged.

```text
...
cudf                      24.10.00a373    cuda12_py312_240923_g8b12cf4e66_373    rapidsai-nightly
cuproj                    24.10.00a75     cuda12_py312_240923_g218cc315_75    file:///tmp/python_channel
cupy                      13.3.0          py312h7d319b9_0    conda-forge
cupy-core                 13.3.0          py312h28031eb_0    conda-forge
cuspatial                 24.10.00a75     cuda12_py312_240923_g218cc315_75    file:///tmp/python_channel
...
fmt                       11.0.2               h434a139_0    conda-forge
...
numba                     0.60.0          py312h83e6fd3_0    conda-forge
numpy                     2.0.2           py312h58c1407_0    conda-forge
...
spdlog                    1.14.1               hed91bc2_1    conda-forge
...
```

([build link](https://github.com/rapidsai/cuspatial/actions/runs/11003344887/job/30553301836?pr=1441))

**Comment by jakirkham** [comment]
Huzzah! 🥳

Thanks for all of the hard work here. Especially pushing through `fmt` / `spdlog` migrations James! 🙏

**Pull request merged**